src/Core/CacheManager.cs

```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Core
+{
+    public class CacheManager
+    {
+        private readonly Func<Task<IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>>>> _asyncDataProvider;
+
+        private IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>> _data;
+
+        private IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>> GetData()
+        {
+            if (_data != null) return _data;
+
+            _data = _asyncDataProvider().Result;
+            return _data;
+        }
+
+        private readonly Timer _refreshTimer;
+
+        public Func<IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>>> DataProvider => GetData;
+
+        public CacheManager(Func<Task<IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>>>> asyncDataProvider, int refreshMilliseconds)
```

Comment:
Constructor should be moved up. As a rule order should be fields, constructor, properties, methods. With exceptions of course...

Reply:
I would suggest `TimeSpan refreshInterval` instead of `int refreshMilliseconds`.